libparsec/crates/client/tests/unit/certif/bootstrap_workspace.rs

```diff
@@ -0,0 +1,824 @@
+                    key_index: 1,
+                    keys_bundle,
+                    keys_bundle_access,
+                }
+            }
+        },
+        {
+            move |req: authenticated_cmds::latest::realm_rename::Req| {
+                p_assert_eq!(req.initial_name_or_fail, true);
+                rep
+            }
+        },
+    );
+
+    let err = ops
+        .bootstrap_workspace(realm_id, &"wksp1".parse().unwrap())
+        .await
+        .unwrap_err();
+
+    assert(err);
+}
+
+#[parsec_test(testbed = "minimal")]
+async fn server_initial_name_already_exists(env: &TestbedEnv) {
+    let (env, realm_id) = env.customize_with_map(|builder| {
+        let realm_id = builder
+            .new_realm("alice")
+            .then_do_initial_key_rotation()
+            .map(|event| event.realm);
+        builder.certificates_storage_fetch_certificates("alice@dev1");
+        builder.rename_realm(realm_id, "wksp1");
+        realm_id
+    });
+    let alice = env.local_device("alice@dev1");
+    let ops = certificates_ops_factory(&env, &alice).await;
+
+    let keys_bundle = env.get_last_realm_keys_bundle(realm_id);
+    let keys_bundle_access = env.get_last_realm_keys_bundle_access_for(realm_id, &alice.user_id());
+    test_register_sequence_of_send_hooks!(
+        &env.discriminant_dir,
+        {
+            move |req: authenticated_cmds::latest::realm_get_keys_bundle::Req| {
+                p_assert_eq!(req.key_index, Some(1));
+                p_assert_eq!(req.realm_id, realm_id);
+
+                authenticated_cmds::latest::realm_get_keys_bundle::Rep::Ok {
+                    key_index: 1,
+                    keys_bundle,
+                    keys_bundle_access,
+                }
+            }
+        },
+        {
+            move |req: authenticated_cmds::latest::realm_rename::Req| {
+                p_assert_eq!(req.initial_name_or_fail, true);
+                authenticated_cmds::latest::realm_rename::Rep::InitialNameAlreadyExists {
+                    last_realm_certificate_timestamp: "2000-01-02T00:00:00Z".parse().unwrap(),
+                }
+            }
+        },
+    );
+
+    let res = ops
+        .bootstrap_workspace(realm_id, &"wksp1".parse().unwrap())
+        .await
+        .unwrap();
+
+    p_assert_matches!(
+        res,
+        CertificateBasedActionOutcome::RemoteIdempotent { certificate_timestamp }
+        if certificate_timestamp == "2000-01-02T00:00:00Z".parse().unwrap()
+    );
+}
+
+#[parsec_test(testbed = "minimal")]
+#[case::invalid_key_index(
+    |env: &TestbedEnv, realm_id, user_id: &UserID| authenticated_cmds::latest::realm_get_keys_bundle::Rep::Ok {
+        key_index: 8,
+        keys_bundle: env.get_last_realm_keys_bundle(realm_id),
+        keys_bundle_access: env.get_last_realm_keys_bundle_access_for(realm_id, user_id),
+    },
+    |err| p_assert_matches!(err, CertifBootstrapWorkspaceError::Offline)
```

Comment:
why `Offline` and not `InvalidKeysBundle` ? (if that's not a bug, we should have a small comment to explain)

Reply:
I don't know.
I think there is an error, because some methods return an offline and other don't return error, but it can be fixed lately